### ISSUE TITLE
A question about the experimental results in the paper

### ISSUE THREAD

**Issue by rrrwwwxxx** [yifanycc/loretta#6]
Hi, thanks for your great work!

I have a quick question about the experimental results reported in your paper.
Are the reported metrics (e.g., accuracy on SST-2) evaluated on the validation set or the test set?

Thanks in advance!

**Comment by yifanycc** [comment]
Hi @rrrwwwxxx,

Thanks for the interest in our paper. The GLUE results are evaluated on the
validation set locally.

Best,
Yifan

**Comment by rrrwwwxxx** [comment]
Thank you for your previous response.

I’ve been working on reproducing the results from your paper, but I’m noticing some discrepancies. My results are generally lower than what’s reported in the paper—some tasks are off by 1-2 points, and others are as much as 10-15 points lower.

Could you share any insights into what might be causing these differences, or any recommendations on how I can improve the results? I’d really appreciate any advice you have!

Best regards

**Comment by yifanycc** [comment]
Sorry for the late reply, I'm busy with relocation these days. If you are
working on the bert-based model. One possible reason is the setup of
trainable parameters. Besides the tensorized adapters, you will need to
activate the last layer (cls head) and the layer norm to be trainable. I
also store some records for the experiments on llama. Here are the logs on
copa, I also have some other records on wandb, I can send them to you if
you need.

**Comment by yifanycc** [comment]
https://drive.google.com/drive/folders/1HSh5QhzI3K3A7qvqVgTGpOXnXR4SngSR?usp=sharing

sorry, forget to attach the link

**Comment by rrrwwwxxx** [comment]
Thank you very much for your patient reply. I recently resumed working on reproducing the RTE task using DeBERTa. Based on your suggestion about activating the last layer (CLS Head) and LayerNorm, I added the following code after "model = get_peft_model(model, peft_config)":
unfrozen_keywords = ['layernorm', 'classifier', 'encoder.layer.11']
for name, param in model.named_parameters():
    if any(k in name.lower() for k in unfrozen_keywords):
        param.requires_grad = True
I also printed the number of trainable parameters in the terminal: Total trainable params: 8341730. I would like to confirm if this is the correct way to implement it. If not, I would really appreciate your guidance.
However, with this setup, the evaluation accuracy (eval_acc) I get on RTE is still only around 61. I'm wondering whether you used the same hyperparameter settings across all datasets (as suggested in the paper/code), or did you tune them separately for each task?

Looking forward to your valuable insights!
Best regards

**Comment by rrrwwwxxx** [comment]
Dear Yifan @yifanycc ,
I'm very interested in your Loretta project and am keen to learn more about its design and implementation. I also wanted to thank you again for your previous help and guidance on the issue; it was very helpful.
I am currently trying to reproduce your Llama experiment results on the COPA dataset. However, I've encountered a subtle but crucial discrepancy in the model's parameter shapes compared to the ones you shared in your W&B logs.
The difference seems to be specifically within the adapter_up modules of both gate_proj and up_proj. In my local run, all adapters (both adapter_down and adapter_up) share the same tensor.factors shape pattern. For example, here are the shapes for gate_proj.adapter_up from my run (at layer 31):
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.bias: torch.Size([64]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.0: torch.Size([1, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.1: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.2: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.3: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.4: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.5: torch.Size([5, 4, 1]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.bias: torch.Size([4096]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.0: torch.Size([1, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.1: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.2: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.3: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.4: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.5: torch.Size([5, 4, 1]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.bias: torch.Size([64]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.0: torch.Size([1, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.1: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.2: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.3: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.4: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.5: torch.Size([5, 4, 1]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.bias: torch.Size([4096]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.0: torch.Size([1, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.1: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.2: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.3: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.4: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.5: torch.Size([5, 4, 1]) parameters

However, in your shared W&B logs, the tensor.factors shapes for the adapter_up module are different, with the 16 and 4 dimensions swapped:
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.bias: torch.Size([64]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.0: torch.Size([1, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.1: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.2: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.3: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.4: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_down.layer.tensor.factors.5: torch.Size([5, 4, 1]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.bias: torch.Size([4096]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.0: torch.Size([1, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.1: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.2: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.3: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.4: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.gate_proj.adapter_up.layer.tensor.factors.5: torch.Size([5, 16, 1]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.bias: torch.Size([64]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.0: torch.Size([1, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.1: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.2: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.3: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.4: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_down.layer.tensor.factors.5: torch.Size([5, 4, 1]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.bias: torch.Size([4096]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.0: torch.Size([1, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.1: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.2: torch.Size([5, 4, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.3: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.4: torch.Size([5, 16, 5]) parameters
base_model.model.model.layers.31.mlp.up_proj.adapter_up.layer.tensor.factors.5: torch.Size([5, 16, 1]) parameters

I suspect this structural difference is the key factor preventing me from reproducing the results. What I find most puzzling, and what I hope you can confirm, is that despite this shape mismatch, the total and trainable parameter counts in my run are exactly identical to your W&B records. This suggests that while the total number of parameters is correct, their architectural distribution seems to differ from your implementation.
As a result, my final experiment results are not as expected. For instance:{ "accuracy": 0.83, "dev_accuracy": 0.91}
I would be extremely grateful for your guidance. Could you please help clarify what might be causing this dimensional swap in the adapter_up factors? For example, could it be a configuration flag I've missed, or perhaps a code version issue? Any advice you could provide to help me resolve this and successfully reproduce the paper's results would be immensely appreciated.

Thank you for your time and for your great contribution to the open-source community.
Best regards

**Comment by yifanycc** [comment]
Hey @rrrwwwxxx ,

I suspect this might be due to a difference in the version of our code. I haven’t tested this specific case before, but I think the structural difference shouldn’t significantly affect performance. Could you please share a complete run log with me, including all the hyperparameter settings (like the lr, rank, etc, perhaps saved from Wandb)? I’d be happy to help you investigate the issue.

Thanks,
Yifan

**Comment by rrrwwwxxx** [comment]
Hi, @yifanycc 
Thank you so much for taking the time to reply and for your patient help!
Since the hyperparameter settings were not saved on wandb, I’ve included the terminal output from my local training process in this .txt file for your reference. 

[output.txt](https://github.com/user-attachments/files/21139604/output.txt)

I really appreciate your continued support and guidance!

Best regards

**Comment by yifanycc** [comment]
Hey @rrrwwwxxx ,

Sorry for the late reply, I'm quiet busy during daytime recently for the summer intern. I just compared your running log with mine and actually I didn't find any major difference... 

I could only say the possible reason could be:
- The  discrepancy in the model's parameter shapes: You could edit the line of code here to make the shape the same as my previous implementation to take a try:
https://github.com/yifanycc/loretta/blob/c66bbb3fe0a717d04dd02e4c5278bf2b61016348/loretta/loretta/utils/tensor_util.py#L70
- **I remember I may load the model in FP32 format at that time.** Maybe the tensor computation part need higher precision for a good performance, I didn't think deep into this point when write the paper.
- Also, I provide some other log for your reference at the google drive 
https://drive.google.com/drive/folders/1HSh5QhzI3K3A7qvqVgTGpOXnXR4SngSR?usp=sharing

**Comment by rrrwwwxxx** [comment]
Hi @yifanycc 
Thank you for your quick reply and guidance!
I followed your suggestions and ran new experiments. First, I loaded the model in full FP32 precision, and the results are as follows:
{"accuracy": 0.82, "dev_accuracy": 0.91}
I adjusted the TT-shape configuration you mentioned, changing the decomposition for 4096 from [16, 16, 16, 4, 4, 4] to [4, 4, 4, 16, 16, 16]. 
However, I observed that this change didn't seem to have any effect. I re-inspected the model's parameter shapes, and they remained the same as before the change. Consequently, the final results were also identical.
I was wondering if you could shed some light on why changing the TT-shape configuration might not be working as expected. Or is there a specific procedure or flag required for the TT-shape configuration to be applied correctly?

Thanks again for your time and help.

Best regards

**Comment by rrrwwwxxx** [comment]
Hi @yifanycc 
Hope you are having a good week. I'm just following up on our previous discussion. First, thank you again for your valuable guidance.
Following your last suggestions, I've made some new progress. I was able to successfully modify the TT-shape configuration and have compared the results under different settings (all in FP32 precision):

With the configuration [16, 16, 16, 4, 4, 4], the result is {"accuracy": 0.82, "dev_accuracy": 0.91}.
When changed to [4, 4, 4, 16, 16, 16], the result became {"accuracy": 0.83, "dev_accuracy": 0.9}.

This result suggests the parameter shape configuration is an important factor for the final outcome.

However, my core issue remains: no matter how I adjust the configuration, I seem to only be able to switch between two symmetric shape patterns:
Pattern 1:
[1, 16, 5]
[5, 16, 5]
[5, 16, 5]
[5, 4, 5]
[5, 4, 5]
[5, 4, 1]
Pattern 2:
[1, 4, 5]
[5, 4, 5]
[5, 4, 5]
[5, 16, 5]
[5, 16, 5]
[5, 16, 1]
Neither of these patterns matches the asymmetric structure in your W&B logs, where adapter_up and adapter_down appear to have different shape patterns. I am still unable to reproduce your exact shape configuration.
Therefore, I would be very grateful if you could offer further guidance. Am I missing a key configuration or a specific implementation detail that enables this asymmetric shape assignment?

I would be immensely grateful if you could spare a moment to offer some suggestions or guidance.
Best regards
